src/modules/systemd/libzbxsystemd.c

```diff
@@ -211,8 +211,8 @@ static int SYSTEMD_UNIT_DISCOVERY(AGENT_REQUEST *request, AGENT_RESULT *result)
       switch (i) {
       case 0:
         // filter by unit type
-        if(NULL != filter || '\0' != *filter)
-          if(0 == systemd_cmptype(value.str, filter))
+        if(NULL != filter)
+           if ('\0' != *filter && (0 == systemd_cmptype(value.str, filter))
```

Comment:
Maybe you should have just replaced `||` with `&&`?
```suggestion
        if(NULL != filter && '\0' != *filter)
          if(0 == systemd_cmptype(value.str, filter))
```

Reply:
Thanks!